main/streams/streams.c

```diff
@@ -1555,6 +1555,84 @@ PHPAPI zend_result _php_stream_copy_to_stream_ex(php_stream *src, php_stream *de
 		return SUCCESS;
 	}
 
+#ifdef HAVE_COPY_FILE_RANGE
+	if (php_stream_is(src, PHP_STREAM_IS_STDIO) &&
+	    php_stream_is(dest, PHP_STREAM_IS_STDIO) &&
+	    src->writepos == src->readpos &&
+	    php_stream_can_cast(src, PHP_STREAM_AS_FD) == SUCCESS &&
+	    php_stream_can_cast(dest, PHP_STREAM_AS_FD) == SUCCESS) {
+		/* both php_stream instances are backed by a file
+		   descriptor, are not filtered and the read buffer is
+		   empty: we can use copy_file_range() */
+
+		int src_fd, dest_fd;
+
+		php_stream_cast(src, PHP_STREAM_AS_FD, (void*)&src_fd, 0);
+		php_stream_cast(dest, PHP_STREAM_AS_FD, (void*)&dest_fd, 0);
+
+		/* clamp to INT_MAX to avoid EOVERFLOW */
+		const size_t cfr_max = MIN(maxlen, (size_t)INT_MAX);
```

Comment:
shouldn't be SSIZE_MAX ?

Reply:
I was not sure what the kernel's real limit was; I remember long ago I read that Linux's limit for all I/O was 2 GB which is INT_MAX. I'll find out!

Reply:
It's ... more complicated than that.
https://github.com/torvalds/linux/blob/v5.17/fs/read_write.c#L1434=
```
	/* Ensure offsets don't wrap. */
	if (pos_in + count < pos_in || pos_out + count < pos_out)
		return -EOVERFLOW;
```
(Note that this code is buggy because the compiler is allowed to optimize this check away.)
(Also note that this check is rather clumsy because the `count` variable is clamped to the remaining size of the file only after this check.)
`SSIZE_MAX` is `2^63-1` on 64 bit CPUs, which means it's safe until you encounter a file of `2^63` bytes, which will not happen for a very loooong time. Should be safe enough to say "copy as much as you can", I guess.